digdag-core/src/main/java/io/digdag/core/agent/ConfigEvalEngine.java

```diff
@@ -51,94 +45,116 @@ private static String readResource(String resourceName)
+            LIBRARY_JS_CONTENTS[i] = new String[] {
+                LIBRARY_JS_RESOURCES[i][0],
+                readResource(LIBRARY_JS_RESOURCES[i][1])
+            };
         }
-
-        RUNTIME_JS_CONTENTS = builder.build();
     }
 
-    private final ObjectMapper jsonMapper;
-    private final NashornScriptEngineFactory jsEngineFactory;
+    private static ImmutableList<String> NO_EVALUATE_PARAMETERS = ImmutableList.of("_do",  "_else_do");
 
-    @Inject
-    public ConfigEvalEngine()
+    private static enum JsEngineType
     {
-        this.jsonMapper = new ObjectMapper();
-        this.jsEngineFactory = new NashornScriptEngineFactory();
+        NASHORN("nashorn"),
+        GRAAL("graal"),
+        NASHORN_GRAAL_CHECK("nashorn-graal-check");
+
+        final String configName;
+
+        JsEngineType(String configName)
+        {
+            this.configName = configName;
+        }
     }
 
-    protected Config eval(Config config, Config params)
-        throws TemplateException
+    private static JsEngineType parseJsEngineType(String type)
     {
-        ObjectNode object = config.convert(ObjectNode.class);
-        ObjectNode built = new Context(params).evalObjectRecursive(object);
-        return config.getFactory().create(built);
+        switch (type) {
+        case "nashorn":
+            return JsEngineType.NASHORN;
+        case "graal":
+            return JsEngineType.GRAAL;
+        case "nashorn-graal-check":
+            return JsEngineType.NASHORN_GRAAL_CHECK;
+        default:
+            throw new ConfigException("Parameter 'eval.js-engine-type' must be either of nashorn, graal, or nashorn-graal-check but got " + type);
+        }
     }
 
-    private Invocable newTemplateInvocable(Config params)
+    @VisibleForTesting
+    static JsEngineType defaultJsEngineType()
     {
-        ScriptEngine jsEngine = jsEngineFactory.getScriptEngine(new String[] {
-            //"--language=es6",  // this is not even accepted with jdk1.8.0_20 and has a bug with jdk1.8.0_51
-            "--no-java",
-            "--no-syntax-extensions",
-            "-timezone=" + params.get("timezone", String.class),
-        });
-        try {
-            for (String runtimeJs : RUNTIME_JS_CONTENTS) {
-                jsEngine.eval(runtimeJs);
-            }
-        }
-        catch (ScriptException | ClassCastException ex) {
-            throw new IllegalStateException("Unexpected script evaluation failure", ex);
+        // Return NASHORN if runtime Java VM is older than 11. Otherwise,
+        // by default, return GRAAL. This is because org.graalvm.js:js depends
+        // on JVMCI (JEP 243) which is available only from Java 11.
+        String javaSpecVer = System.getProperty("java.specification.version");
+        int major = Integer.parseInt(javaSpecVer.split("[^0-9]")[0]);
+        if (major < 11) {
+            return JsEngineType.NASHORN;
         }
-        return (Invocable) jsEngine;
+        return JsEngineType.GRAAL;
```

Comment:
If we change the default JS engine based on Java version, what about logging to inform it?